executor/sort.go

```diff
@@ -79,20 +126,74 @@ func (e *SortExec) Next(ctx context.Context, req *chunk.Chunk) error {
 		if err != nil {
 			return err
 		}
-		e.initPointers()
-		e.initCompareFuncs()
-		e.buildKeyColumns()
-		sort.Slice(e.rowPtrs, e.keyColumnsLess)
-		e.fetched = true
+		if e.alreadySpilled() {
+			err = e.externalSorting()
+			if err != nil {
+				return err
+			}
+			e.fetched = true
+		} else {
+			e.initPointers()
+			e.initCompareFuncs()
+			e.buildKeyColumns()
+			sort.Slice(e.rowPtrs, e.keyColumnsLess)
+			e.fetched = true
+		}
 	}
-	for !req.IsFull() && e.Idx < len(e.rowPtrs) {
-		rowPtr := e.rowPtrs[e.Idx]
-		req.AppendRow(e.rowChunks.GetRow(rowPtr))
-		e.Idx++
+
+	if e.alreadySpilled() {
+		for !req.IsFull() && e.Idx < len(e.finalRowPtrs) {
+			rowPtr := e.finalRowPtrs[e.Idx]
+			row, err := e.finalChunksInDisk.GetRow(rowPtr)
+			if err != nil {
+				return err
+			}
+			req.AppendRow(row)
+			e.Idx++
+		}
+	} else {
+		for !req.IsFull() && e.Idx < len(e.rowPtrs) {
+			rowPtr := e.rowPtrs[e.Idx]
+			req.AppendRow(e.rowChunks.GetRow(rowPtr))
+			e.Idx++
+		}
 	}
 	return nil
 }
 
+func (e *SortExec) externalSorting() (err error) {
+	e.initCompareFuncs()
+	e.buildKeyColumns()
+	e.rowPtrsInDisk = e.initPointersForListInDisk(e.rowChunksInDisk)
+	// partition sort
+	// Now only have one partition.
+	// The partition will be adjusted in the next pr.
+	err = e.readPartition(e.rowChunksInDisk, e.rowPtrsInDisk)
+	if err != nil {
+		return err
+	}
+	e.initPointers()
+	sort.Slice(e.rowPtrs, e.keyColumnsLess)
+	listInDisk, err := e.spillToDiskByRowPtr()
+	if err != nil {
+		return err
+	}
+	e.partitionList = append(e.partitionList, listInDisk)
+	e.partitionRowPtrs = append(e.partitionRowPtrs, e.initPointersForListInDisk(listInDisk))
+	// merge sort
+	// merge sort will be implemented in the next pr.
+	// Now it only read data from partition and restore again.
+	err = e.readPartition(e.partitionList[0], e.partitionRowPtrs[0])
+	if err != nil {
+		return err
+	}
+	e.initPointers()
```

Comment:
each time we call `initPointers`, it consumes some memory into `memTracker`, we should find a way to release it.